### PR TITLE
CustomSelectControl V2: fix labelling with a visually hidden label

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Part of #55023

Requires #63169 

Fix a bug in the V2 implementation where the component wouldn't be labelled property to assistive technology when the label is hidden visually via the `hideLabelFromVision` prop.

## Why?

Making sure that controls are properly labelled is essential to ensure a good experience for all users.

## How?

I refactored the code and made it so that the internal implementation of `CustomSelectControlV2` always relies on the `Ariakit.SelectLabel` component even when the `hideLabelFromVision` is `true`, which takes care of setting the correct attributes necessary for the control labeling. I was able to do so by leveraging the `render` prop ([see docs](https://ariakit.org/guide/composition#composing-with-custom-components)).

## Testing Instructions

- Make sure that the newly added unit tests pass (and that the newly added tests for the V2 version don't pass on `trunk`)
- Load Storybook examples, play with the `hideLabelFromVision` prop:
  - Make sure that the component looks and behaves like on `trunk`
  - Inspect the DOM and check that the button with `role="combobox"` is correctly labeled by the hidden `label`

### CODE REVIEW

**Review by mirka — COMMENTED** [review]
I feel like the component should be wrapped in a `BaseControl` (for consistency with other controls), or at the very least use `BaseControl.VisualLabel` instead of `Styled.SelectLabel`. Are there some considerations I'm missing?

**Comment by ciampo** [comment]
> Are there some considerations I'm missing?

Not really, my initial fix focused on using the components that were already being used (and that made me miss the `TODO` inline comment). I'll look into using as much of `BaseControl` as possible

**Comment by ciampo** [comment]
@mirka since I don't think it would be possible to use `<BaseControl />` while also using `Ariakit.SelectLabel`, I opted for your second-best strategy of keeping most code as-is and swapping `Styled.SelectLabel` with `BaseControl.VisualLabel`.

I also had to allow `BaseControl.VisualLabel` to forward ref in order for it work correctly with `Ariakit.SelectLabel` (since ariakit [needs to grab a reference to the label element](https://github.com/ariakit/ariakit/blob/main/packages/ariakit-react-core/src/select/select-label.tsx#L62)).

The only visible difference is that the label now renders a tiny gap above itself, due to the fact that `BaseControl.VisualLabel` has `display: inline-block`, which doesn't trim the whitespace. That is an inconsistency across 	`BaseControl`-based components, though:
- components using "pure" `BaseControl` show the extra white space above label
- components using only `BaseControl.VisualLabel` may or may not show the white space based on a bunch of factors. The most common one is whether the label wrapper element is a flex or grid parent (which trims whitespace automatically).

File: packages/components/src/base-control/stories/index.story.tsx
Comment (ciampo) [inline packages/components/src/base-control/stories/index.story.tsx]: Well, it looks like the way to fix it was to allow ref forwarding 🤷‍♂️

File: packages/components/src/custom-select-control-v2/custom-select.tsx
Comment (ciampo) [inline packages/components/src/custom-select-control-v2/custom-select.tsx:124]: Using the `label` HTML element is not correct in this case, as also confirmed by the [ariakit docs](https://github.com/ariakit/ariakit/blob/main/packages/ariakit-react-core/src/select/select-label.tsx#L19-L22):

> Since it's not a native select element, we can't use the native label element.

**Comment by ciampo** [comment]
I extracted the `BaseControl.VisualLabel` changes to a separate PR: #63169

Once merged, I'll rebase and remove those same changes from this PR.

**Comment by mirka** [comment]
> * components using "pure" `BaseControl` show the extra white space above label
> * components using only `BaseControl.VisualLabel` may or may not show the white space based on a bunch of factors. The most common one is whether the label wrapper element is a flex or grid parent (which trims whitespace automatically).

Furthermore, this visual difference won't manifest in a wp-admin context because `common.css` has `body { line-height: 1.4em; }`, normalizing the line-height of `.components-base-control__field`. Do you think it's worth adding an explicit line-height to `components-base-control__field`? 🤔 Could be risky.

File: packages/components/src/custom-select-control-v2/custom-select.tsx
Comment (mirka) [reply packages/components/src/custom-select-control-v2/custom-select.tsx:124]: Interesting! I played around and confirmed that although a native `label` will work as expected in Chrome, it behaves differently in Firefox and Safari. Clicking on the label not only focuses but clicks the "select button", expanding the dropdown. So yeah, I guess it does need special handling 🫤

**Comment by ciampo** [comment]
> Do you think it's worth adding an explicit line-height to components-base-control__field? 🤔 Could be risky.

I agree that it would be risky — let's leave as-is. In the long term, I'd like to re-think `BaseControl` to enable more flexibility for which `<Label />` component to render to better composition with other components, and to handle out of the box [composed inputs](https://github.com/WordPress/gutenberg/issues/42630).

I merged https://github.com/WordPress/gutenberg/pull/63169 and rebased this PR. It should be ready for a last round of review

File: packages/components/src/custom-select-control-v2/custom-select.tsx
Comment (ciampo) [reply packages/components/src/custom-select-control-v2/custom-select.tsx:124]: Yup, that's part of what `Ariakit.SelectLabel` does under the hood

File: packages/components/src/custom-select-control-v2/test/index.tsx
Comment (tyxla) [inline packages/components/src/custom-select-control-v2/test/index.tsx:440]: Thanks for adding a test 🙌

**Review by tyxla — COMMENTED** [review]
Went through a round of testing on this and it tested well according to the test instructions. 

I have nothing to add on the code, IMHO it's good to go, but I'll let @mirka approve it since she had some feedback before.

**Review by mirka — APPROVED** [review]
Perfect, thank you!

**Pull request merged**
